ingest/usecase/plugins/orderbook/claimbot/plugin.go

```diff
@@ -0,0 +1,160 @@
+package claimbot
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/osmosis-labs/sqs/domain"
+	"github.com/osmosis-labs/sqs/domain/keyring"
+	"github.com/osmosis-labs/sqs/domain/mvc"
+	orderbookdomain "github.com/osmosis-labs/sqs/domain/orderbook"
+	orderbookgrpcclientdomain "github.com/osmosis-labs/sqs/domain/orderbook/grpcclient"
+	"github.com/osmosis-labs/sqs/domain/slices"
+	"github.com/osmosis-labs/sqs/log"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+
+	"go.opentelemetry.io/otel"
+	"go.uber.org/zap"
+)
+
+// claimbot is a claim bot that processes and claims eligible orderbook orders at the end of each block.
+// Claimable orders are determined based on order filled percentage that is handled with fillThreshold package level variable.
+type claimbot struct {
+	config     *Config
+	atomicBool atomic.Bool
+}
+
+var _ domain.EndBlockProcessPlugin = &claimbot{}
+
+const (
+	tracerName = "sqs-orderbook-claimbot"
+)
+
+var (
+	tracer                 = otel.Tracer(tracerName)
+	fillThreshold          = osmomath.MustNewDecFromStr("0.98")
+	blockInclusionWaitTime = 5 * time.Second
+)
+
+// maxBatchOfClaimableOrders is the maximum number of claimable orders
+// that can be processed in a single batch.
+const maxBatchOfClaimableOrders = 100
+
+// New creates and returns a new claimbot instance.
+func New(
+	keyring keyring.Keyring,
+	orderbookusecase mvc.OrderBookUsecase,
+	poolsUsecase mvc.PoolsUsecase,
+	orderbookRepository orderbookdomain.OrderBookRepository,
+	orderBookClient orderbookgrpcclientdomain.OrderBookClient,
+	logger log.Logger,
+	chainGRPCGatewayEndpoint string,
+	chainID string,
+) (*claimbot, error) {
+	config, err := NewConfig(keyring, orderbookusecase, poolsUsecase, orderbookRepository, orderBookClient, logger, chainGRPCGatewayEndpoint, chainID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create config: %w", err)
+	}
+
+	return &claimbot{
+		config:     config,
+		atomicBool: atomic.Bool{},
+	}, nil
+}
+
+// ProcessEndBlock implements domain.EndBlockProcessPlugin.
+// This method is called at the end of each block to process and claim eligible orderbook orders.
+// ProcessEndBlock implements domain.EndBlockProcessPlugin.
+func (o *claimbot) ProcessEndBlock(ctx context.Context, blockHeight uint64, metadata domain.BlockPoolMetadata) error {
+	ctx, span := tracer.Start(ctx, "orderbooktFillerIngestPlugin.ProcessEndBlock")
+	defer span.End()
+
+	// For simplicity, we allow only one block to be processed at a time.
+	// This may be relaxed in the future.
+	if !o.atomicBool.CompareAndSwap(false, true) {
+		o.config.Logger.Info("already in progress", zap.Uint64("block_height", blockHeight))
+		return nil
+	}
+	defer o.atomicBool.Store(false)
+
+	orderbooks, err := getOrderbooks(o.config.PoolsUseCase, blockHeight, metadata)
+	if err != nil {
+		return err
+	}
```

Comment:
Instead of having this as a function and propagating pools use case as an arg, why not make this be a method on the pools use case?

Reply:
My attempt to address this comment within reasonable timeframe did not produced satifying result. The problem I faced was testing. If we would end up moving this func as a method to usecase end result might look like following:

```
func (p *poolsUseCase) GetBlockCanonicalOrderbookPools(metadata domain.BlockPoolMetadata) ([]domain.CanonicalOrderBooksResult, error) {
	orderbooks, err := p.GetAllCanonicalOrderbookPoolIDs()
	if err != nil {
		return nil, err
	}

	var result []domain.CanonicalOrderBooksResult
	for _, orderbook := range orderbooks {
		if _, ok := metadata.PoolIDs[orderbook.PoolID]; ok {
			result = append(result, orderbook)
		}
	}
	return result, nil
}
```

But now, we need to either mock `p.GetAllCanonicalOrderbookPoolIDs` or setup in a proper way so expected pool IDs are returned. However in order to cover all scenarios as defined in the [tests](https://github.com/osmosis-labs/sqs/pull/524/files#diff-c34124bfc79d562d366617d7fea859dba1f269644011ff50b798c888272aab10R13) would require a lot of boilerplate as for example [here](https://github.com/osmosis-labs/sqs/blob/BE-586/pools/usecase/pools_usecase_test.go) and we still would not be able to test error case. I think this case is a good example when current design starts limitting us and presents a good opportunity rethink it.

Reply:
The linked example looks much cleaner IMO, and this is how I would expect the API to look like if minimizing code smells

I don't understand what the problem with mocking out `poolsUseCase` was. (the link did not point me to anything specific btw)

Happy to not block this PR on this but a task would be greatly appreciated